File: apps/ios/FloentlyShared/Sources/FloentlyShared/Auth/FloentlyAuthService.swift
import Foundation

public final class FloentlyAuthService {
    private let api: FloentlyAPIClient
    private let store: FloentlySessionStore

    public init(api: FloentlyAPIClient, store: FloentlySessionStore) {
        self.api = api
        self.store = store
    }

    public func login(email: String, password: String) async throws -> FloentlyAuthSession {
        let session = try await api.post(
            "/api/v1/auth/login/password",
            body: PasswordLoginRequest(email: email, password: password),
            as: FloentlyAuthSession.self
        )
        store.save(session)
        return session
    }

    public func register(email: String, password: String, name: String?) async throws -> FloentlyAuthSession {
        let session = try await api.post(
            "/api/v1/auth/register/password",
            body: PasswordRegisterRequest(email: email, password: password, name: name, captchaToken: nil),
            as: FloentlyAuthSession.self
        )
        store.save(session)
        return session
    }

    public func requestPasswordReset(email: String) async throws {
        _ = try await api.post(
            "/api/v1/auth/password-reset/request",
            body: PasswordResetRequest(email: email),
            as: FloentlyEmpty.self
        )
    }

    public func confirmPasswordReset(token: String, password: String) async throws {
        _ = try await api.post(
            "/api/v1/auth/password-reset/confirm",
            body: PasswordResetConfirmRequest(token: token, password: password),
            as: FloentlyEmpty.self
        )
    }

    public func restoreSession() async throws -> FloentlyAuthSession {
        let session = try await api.get("/api/v1/auth/session", as: FloentlyAuthSession.self)
        store.save(session)
        return session
    }

    public func logout() async {
        _ = try? await api.post(
            "/api/v1/auth/logout",
            body: LogoutRequest(refreshToken: store.session?.refreshToken),
            as: FloentlyEmpty.self
        )
        store.clear()
    }
}

File: apps/android/shared/src/main/java/com/floently/shared/auth/FloentlyAuthService.kt
package com.floently.shared.auth

import com.floently.shared.api.FloentlyApiClient
import org.json.JSONObject

class FloentlyAuthService(
    private val api: FloentlyApiClient,
    private val store: FloentlySecureSessionStore
) {
    suspend fun login(email: String, password: String): FloentlyAuthSession {
        val response = api.post(
            "/api/v1/auth/login/password",
            JSONObject().put("email", email).put("password", password)
        )
        return authSessionFromJson(response).also { store.save(it) }
    }

    suspend fun register(email: String, password: String, name: String?): FloentlyAuthSession {
        val body = JSONObject()
            .put("email", email)
            .put("password", password)

        if (!name.isNullOrBlank()) body.put("name", name)

        val response = api.post("/api/v1/auth/register/password", body)
        return authSessionFromJson(response).also { store.save(it) }
    }

    suspend fun requestPasswordReset(email: String) {
        api.post("/api/v1/auth/password-reset/request", JSONObject().put("email", email))
    }

    suspend fun confirmPasswordReset(token: String, password: String) {
        api.post(
            "/api/v1/auth/password-reset/confirm",
            JSONObject().put("token", token).put("password", password)
        )
    }

    suspend fun restoreSession(): FloentlyAuthSession {
        val response = api.get("/api/v1/auth/session")
        return authSessionFromJson(response).also { store.save(it) }
    }

    suspend fun logout() {
        runCatching {
            api.post(
                "/api/v1/auth/logout",
                JSONObject().put("refresh_token", store.session?.refreshToken)
            )
        }
        store.clear()
    }
}
